### PR TITLE
Move logs to user storage + Improvements to force close logs

### PR DESCRIPTION
**Pull request by amstone326**
Changes to what we do with force closure reports:
-If submission of a force close fails on initial attempt, save it to storage (if possible, e.g. if user is logged in) so we can try again later during a normal `LogSubmissionTask`
-Make a dedicated `ForceCloseLogEntry` class that stores a lot more info than other log entries have (domain, app version, readable session string, restorable serialized session string, etc.), and start submitting force close logs in this format (but for now, also continues sending them in the old format, so that HQ can still process them until the new format is being handled)

And a change to how we handle all logs: When there is an active session, save them to user storage, rather than global storage

### CODE REVIEW

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                });
+            }
+        });
+    }
+
+    @Override
+    public void writeToDeviceReport(XmlSerializer serializer) throws IOException {
+        this.serializer = serializer;
+
+        serializer.startTag(DeviceReportWriter.XMLNS, "force_close_subreport");
```
Comment (ctsims) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: This means we're going to start missing these in the short term, right? Until we get HQ to start parsing these?

File: app/src/org/commcare/android/session/DevSessionRestorer.java [hunk 2]
```diff
@@ -108,26 +108,36 @@ public static void saveSessionToPrefs() {
...
         }
-
         SharedPreferences prefs = ccApp.getAppPreferences();
+        String serializedSession = getSerializedSessionString();
+        if (!"".equals(serializedSession)) {
+            prefs.edit().putString(CommCarePreferences.CURRENT_SESSION, serializedSession).commit();
+        }
+    }
+
+    public static String getSerializedSessionString() {
```
Comment (ctsims) [inline app/src/org/commcare/android/session/DevSessionRestorer.java]: How similar is this method to your utility set above?

**Comment by ctsims** [comment]
minor comment and question, otherwise good to pull once you're checked off on "don't pull"

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java (hunk 1 above)
Comment (amstone326) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: @ctsims Yea that's a good point. Would probably need to wait to pull this til that happens? Or I could add something in temporarily that puts them in the normal device logs as well. Not sure what makes the most sense there.

File: app/src/org/commcare/android/session/DevSessionRestorer.java (hunk 2 above)
Comment (amstone326) [inline app/src/org/commcare/android/session/DevSessionRestorer.java]: @ctsims Sorry, not sure what you're asking about. What utility above?

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java (hunk 1 above)
Comment (ctsims) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: I think adding it to both when we log is probably the easiest short-term solution if it's not too onerous.

File: app/src/org/commcare/android/session/DevSessionRestorer.java (hunk 2 above)
Comment (ctsims) [inline app/src/org/commcare/android/session/DevSessionRestorer.java]: Ah, I think I was confusing what you're calling the "readable session string" above, and the "Serialized Session String" here.

File: app/src/org/commcare/android/logging/ForceCloseLogger.java [hunk 3]
```diff
@@ -112,12 +117,31 @@ public String getFilename() {
...
+        }
+    }
+
+    private static String getSubmissionUri() {
+        try {
+            SharedPreferences settings = CommCareApplication._().getCurrentApp().getAppPreferences();
+            return settings.getString(CommCarePreferences.PREFS_SUBMISSION_URL_KEY,
+                    CommCareApplication._().getString(R.string.PostURL));
+        } catch (Exception e) {
+            return CommCareApplication._().getString(R.string.PostURL);
```
Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: Are we trying to catch NPEs when there isn't an app seated? I would prefer explicit `null` checks or catching `NullPointerException` over the general `Exception`

File: app/src/org/commcare/android/logging/ForceCloseLogger.java [hunk 4]
```diff
@@ -112,12 +117,31 @@ public String getFilename() {
...
+    private static void writeErrorToStorage(ForceCloseLogEntry entry) {
+        if (logStorage != null) {
+            logStorage.write(entry);
+        }
+    }
+
+    private static String getSubmissionUri() {
+        try {
+            SharedPreferences settings = CommCareApplication._().getCurrentApp().getAppPreferences();
+            return settings.getString(CommCarePreferences.PREFS_SUBMISSION_URL_KEY,
```
Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: There are a lot of places that have these two lines of code, might be good to eventually replace them with this method.

File: app/src/org/commcare/CommCareApplication.java [hunk 5]
```diff
@@ -947,7 +957,16 @@ public void onServiceConnected(ComponentName className, IBinder service) {
                         }
                     }
 
-                    XPathErrorLogger.registerStorage(getUserStorage(XPathErrorEntry.STORAGE_KEY, XPathErrorEntry.class));
+                    // Switch the regular and forece close loggers over to using user storage while
+                    // there is a session
+                    Logger.registerLogger(new AndroidLogger(getUserStorage(AndroidLogEntry.STORAGE_KEY,
+                            AndroidLogEntry.class)));
+                    ForceCloseLogger.registerStorage(getUserStorage(ForceCloseLogEntry.STORAGE_KEY,
+                            ForceCloseLogEntry.class));
```
Comment (phillipm) [inline app/src/org/commcare/CommCareApplication.java]: These logger registrations could be lifted into a helper like `void setupLoggers(boolean isUserStorageAvailable)`

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import org.javarosa.core.model.utils.DateUtils;
+import org.javarosa.core.services.storage.EntityFilter;
+import org.javarosa.core.util.SortedIntSet;
+import org.xmlpull.v1.XmlSerializer;
+
+import java.io.IOException;
+import java.util.Hashtable;
+
+/**
+ * Convert xpath error logs to xml.
```
Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: needs to be updated

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            writeText("session_readable", forceCloseEntry.getReadableSession());
+            writeText("session_serialized", forceCloseEntry.getSerializedSessionString());
+        } catch (Exception e) {
+            e.printStackTrace();
+        } finally {
+            serializer.endTag(DeviceReportWriter.XMLNS, "force_close");
+        }
+    }
+
+    private void writeText(String element, String text)
```
Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: Duplicated in `XPathErrorSerializer`

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                        return IDs.contains(id) ? PREFILTER_INCLUDE : PREFILTER_EXCLUDE;
+                    }
+
+                    public boolean matches(LogEntry e) {
+                        throw new RuntimeException("can't happen");
+                    }
+                });
+            }
+        });
+    }
```
Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: Duplicated in `XPathErrorSerializer`. Might be worth making them both subclasses of the same base class or having some static purge setup methods they can both call

File: app/src/org/commcare/android/logging/ForceCloseLogger.java [hunk 9]
```diff
@@ -26,61 +26,65 @@
...
-            report = null;
-        }
-
+    /**
+     * Attempts to send a force close report for the given exception to the server
+     * immediately. If this fails, instead writes the exception to storage as a ForceCloseLogEntry,
+     * so that we can attempt to send it again later during a normal LogSubmissionTask
+     */
+    private static void sendToServerOrStore(Throwable exception) {
+        ByteArrayOutputStream streamToWriteErrorTo = new ByteArrayOutputStream();
```
Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: This stream is never closed

Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: Nevermind, it is in `DeviceReportWriter.write()`

Comment (phillipm) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: Errr, unless the `DeviceReportWriter` constructor throws an `IOException`, then it isn't closed...

File: app/src/org/commcare/logging/AndroidLogEntry.java [hunk 10]
```diff
@@ -37,30 +33,24 @@ public AndroidLogEntry() {
     }
 
     public AndroidLogEntry(String type, String message, Date date) {
-        this.type = type;
-        this.message = message;
-        this.date = date;
+        super(type, message, date);
```
Comment (phillipm) [inline app/src/org/commcare/logging/AndroidLogEntry.java]: :+1:

File: app/src/org/commcare/logging/AndroidLogSerializer.java [hunk 11]
```diff
@@ -78,11 +79,12 @@ public void writeToDeviceReport(XmlSerializer serializer) throws IOException {
         serializer.startTag(DeviceReportWriter.XMLNS, "log_subreport");
 
         try {
-            if (isSingleLog) {
-                serializeLog(entry);
-            } else {
-                if (Logger._() != null) {
-                    Logger._().serializeLogs(this);
```
Comment (phillipm) [inline app/src/org/commcare/logging/AndroidLogSerializer.java]: Seems like `serializeLogs` is not called anymore. Maybe you can remove the implementations from `AndroidLogger` then?

File: app/src/org/commcare/preferences/DevSessionRestorer.java [hunk 12]
```diff
@@ -108,27 +108,37 @@ public static void saveSessionToPrefs() {
...
+        String serializedSession = getSerializedSessionString();
+        String formEntrySession = FormEntryActivity.getFormEntrySessionString();
+
+        ccApp.getAppPreferences().edit()
+                .putString(CommCarePreferences.CURRENT_SESSION, serializedSession)
+                .putString(CommCarePreferences.CURRENT_FORM_ENTRY_SESSION, formEntrySession)
+                .apply();
+    }
+
+    public static String getSerializedSessionString() {
```
Comment (phillipm) [inline app/src/org/commcare/preferences/DevSessionRestorer.java]: :+1:

**Comment by phillipm** [comment]
Thanks for doing this, it was much needed!
Left some comments that should be addressed, then good to go!

File: app/src/org/commcare/android/logging/ForceCloseLogger.java (hunk 3 above)
Comment (amstone326) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: I don't know for sure what it's trying to catch here, this was just refactored from this: https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/tasks/ExceptionReporting.java#L74-L80. I can't really imagine what else it would be protecting against besides the current app being null though, so feel comfortable changing it to that.

File: app/src/org/commcare/android/logging/ForceCloseLogger.java (hunk 4 above)
Comment (amstone326) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: Will probably leave that for another PR

File: app/src/org/commcare/android/logging/ForceCloseLogSerializer.java (hunk 6 above)
Comment (amstone326) [inline app/src/org/commcare/android/logging/ForceCloseLogSerializer.java]: whoops

File: app/src/org/commcare/android/logging/ForceCloseLogger.java (hunk 9 above)
Comment (amstone326) [inline app/src/org/commcare/android/logging/ForceCloseLogger.java]: Hm, confused what you mean on this for a couple of reasons. The `finally` block of `DeviceReportWriter.write()` does call `os.close()`, so not sure what you mean by the last comment. But is also worth noting that the default implementations of `close()` for both `OutputStream` and `ByteArrayOutputStream` do nothing, so unless we introduce an override, it's kind of irrelevant. I'm not sure what an overridden version of that method would do, but if you have an idea definitely share.

File: app/src/org/commcare/logging/AndroidLogSerializer.java (hunk 11 above)
Comment (amstone326) [inline app/src/org/commcare/logging/AndroidLogSerializer.java]: Hm, I can't do that unless I make `AndroidLogger` abstract because it inherits those methods from `ILogger`, which is an interface. And it does look like j2me still makes use of the `serializeLogs` methods in `ILogger`, so those methods can't be totally removed. I could make the `AndroidLogger` implementations return `null` though, to indicate that they're unused. Not sure how I feel about that pattern, but open to it.

**Comment by amstone326** [comment]
@phillipm thanks for the feedback. Implemented all suggestions except for a couple that I left questions on.
